### PR TITLE
Cyclops synchs

### PR DESCRIPTION
**Pull request by Ghaarg**
Cyclops synchs for:
 - Decoy launcher
 - Silent Running
 - Sonar (for some reason the invoke repeating does not work, therefore the workaround for every ping)
 - Cyclops horn
 - Cyclops shield
 - Equipment for the cyclops will not break the save anymore

### CODE REVIEW

**Comment by killzoms** [comment]
GG

File: NitroxClient/Communication/Packets/Processors/CyclopsSonarPingProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            this.packetSender = packetSender;
+        }
+
+        public override void Process(CyclopsSonarPing sonarPacket)
+        {
+            GameObject cyclops = GuidHelper.RequireObjectFrom(sonarPacket.Guid);
+            CyclopsSonarButton sonar = cyclops.RequireComponentInChildren<CyclopsSonarButton>();
+            using (packetSender.Suppress<CyclopsSonarPing>())
+            {
+                MethodInfo sonarPingInfo = sonar.GetType().GetMethod("SonarPing", BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (Sunrunner37) [inline NitroxClient/Communication/Packets/Processors/CyclopsSonarPingProcessor.cs]: I think we can use the helper method `ReflectionCall` here.

File: NitroxClient/Communication/Packets/Processors/CyclopsActivateSonarProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        public override void Process(CyclopsActivateSonar sonarPacket)
+        {            
+            GameObject cyclops = GuidHelper.RequireObjectFrom(sonarPacket.Guid);
+            CyclopsSonarButton sonar = cyclops.RequireComponentInChildren<CyclopsSonarButton>();
+            using (packetSender.Suppress<CyclopsActivateSonar>())
+            {
+                // At this moment the code is "non functional" as for some reason changing the sprite will never happen
+                // Also setting sonar as active will never work
+                //Log.Debug("Process sonar click guid: " + sonar.gameObject.GetGuid().ToString());
```
Comment (Sunrunner37) [inline NitroxClient/Communication/Packets/Processors/CyclopsActivateSonarProcessor.cs]: Can we remove the commented out code?  (here and below)

File: NitroxClient/Communication/Packets/Processors/CyclopsChangeSilentRunningProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        public override void Process(CyclopsChangeSilentRunning packet)
+        {
+            GameObject cyclops = GuidHelper.RequireObjectFrom(packet.Guid);
+            CyclopsSilentRunningAbilityButton ability = cyclops.RequireComponentInChildren<CyclopsSilentRunningAbilityButton>();
+
+            using (packetSender.Suppress<CyclopsChangeSilentRunning>())
+            {
+                ability.GetType().GetField("active", BindingFlags.NonPublic | BindingFlags.Instance).SetValue(ability, packet.IsOn);
```
Comment (Sunrunner37) [inline NitroxClient/Communication/Packets/Processors/CyclopsChangeSilentRunningProcessor.cs]: We can use the helper `ReflectionSet` for this.

File: NitroxPatcher/Patches/CyclopsShieldButton_OnClick_Patch.cs [hunk 4]
```diff
@@ -1,26 +1,67 @@
...
+                instructionList.Insert(0, new CodeInstruction(OpCodes.Nop));
+            }
+            return instructionList;
+        }
+
         public override void Patch(HarmonyInstance harmony)
         {
             PatchPostfix(harmony, TARGET_METHOD);
-        }
+            PatchTranspiler(harmony, TARGET_METHOD);
```
Comment (Sunrunner37) [inline NitroxPatcher/Patches/CyclopsShieldButton_OnClick_Patch.cs]: You may want to use PatchMultiple so they are applied at the same time.  I'm not sure of the direct result of running one right after the other.

File: NitroxPatcher/Patches/CyclopsSonarButton_SonarPing_Patch.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        public static readonly Type TARGET_CLASS = typeof(CyclopsSonarButton);
+        public static readonly MethodInfo TARGET_METHOD = TARGET_CLASS.GetMethod("SonarPing", BindingFlags.NonPublic | BindingFlags.Instance);
+        public static readonly OpCode JUMP_TARGET_CODE = OpCodes.Ldsfld;
+        public static readonly FieldInfo JUMP_TARGET_FIELD = typeof(SNCameraRoot).GetField("main", BindingFlags.Public | BindingFlags.Static);
+
+        public static void Postfix(CyclopsSonarButton __instance)
+        {
+            string guid = GuidHelper.GetGuid(__instance.subRoot.gameObject);
+            //bool activeSonar = Traverse.Create(__instance).Field("sonarActive").GetValue<bool>();
```
Comment (Sunrunner37) [inline NitroxPatcher/Patches/CyclopsSonarButton_SonarPing_Patch.cs]: dead code

File: NitroxPatcher/Patches/CyclopsSonarButton_SonarPing_Patch.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+namespace NitroxPatcher.Patches
+{
+    class CyclopsSonarButton_SonarPing_Patch : NitroxPatch
+    {
+        public static readonly Type TARGET_CLASS = typeof(CyclopsSonarButton);
+        public static readonly MethodInfo TARGET_METHOD = TARGET_CLASS.GetMethod("SonarPing", BindingFlags.NonPublic | BindingFlags.Instance);
+        public static readonly OpCode JUMP_TARGET_CODE = OpCodes.Ldsfld;
+        public static readonly FieldInfo JUMP_TARGET_FIELD = typeof(SNCameraRoot).GetField("main", BindingFlags.Public | BindingFlags.Static);
+
+        public static void Postfix(CyclopsSonarButton __instance)
```
Comment (Sunrunner37) [inline NitroxPatcher/Patches/CyclopsSonarButton_SonarPing_Patch.cs]: Can we add a description above each patch class that tells what it is doing?

Comment (Ghaarg) [reply NitroxPatcher/Patches/CyclopsSonarButton_SonarPing_Patch.cs]: I commented the specific methods. I hope that´s what you wanted.

**Pull request merged**